### PR TITLE
azure image builder with azure windows baseline

### PR DESCRIPTION
**Pull request by mgreenegit**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by mgreenegit** [comment]
#sign-off

File: demos/imagebuilder-windowsbaseline/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+## Resources
+
+The following resources are created by this template:
+
+- 1 user-assigned managed identity for running a deployment script
+- 1 role definition and 1 role assignment to limit access of the new identity
+- 1 Azure Image Gallery with 1 image and 1 template
+- 1 deployment script to trigger the build of the custom image in Azure Image Builder
+
+<img src="images/resources.png" />
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/README.md:49]: the images aren't included in the sample

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+param customizerScriptName string = 'scripts/runScript.ps1'
+
+@description('Name of the user-assigned managed identity used by Azure Image Builder template, and for triggering the Azure Image Builder build at the end of the deployment')
+param identityName string = substring('ImageGallery_${guid(resourceGroup().id)}', 0, 21)
+
+@description('Permissions to allow for the user-assigned managed identity.')
+param roleDefinitionName string = guid(resourceGroup().id)
+
+@description('Name of the role to grant for the user-assigned managed identity.')
+param roleAssignmentName string = guid(resourceGroup().id)
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:21]: this needs to incorporate the principal, role and scope - and you'll want to make it a variable so users can't change it (and fail)

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+@description('Name of the template to create in Azure Image Builder.')
+param imageTemplateName string = 'Win2019_AzureWindowsBaseline_Template'
+
+@description('Name of the custom iamge to create and distribute using Azure Image Builder.')
+param runOutputName string = 'Win2019_AzureWindowsBaseline_CustomImage'
+
+@description('List the regions in Azure where you would like to replicate the custom image after it is created.')
+param replicationRegions array = [
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:41]: us only?

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+resource roleDefinitionName_resource 'Microsoft.Authorization/roleDefinitions@2018-07-01' = {
+  name: roleDefinitionName
+  properties: {
+    roleName: roleDefinitionName
+    description: 'Used for AIB template and ARM deployment script that runs AIB build'
+    type: 'customRole'
+    isCustom: true
+    permissions: [
+      {
+        actions: [
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:69]: are we creating a built-in definition for this?  (eventually?)

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  'westus2'
+  'westus3'
+  'northcentralus'
+  'southcentralus'
+  'westcentralus'
+]
+
+var customizerScriptUri = uri(_artifactsLocation, '${customizerScriptName}${_artifactsLocationSasToken}')
+
+resource identityName_resource 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' = {
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:55]: will need to clean up the decompile lint (see pipeline results)

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+      resourceGroup().id
+    ]
+  }
+}
+
+resource roleAssignmentName_resource 'Microsoft.Authorization/roleAssignments@2021-04-01-preview' = {
+  name: roleAssignmentName
+  properties: {
+    roleDefinitionId: roleDefinitionName_resource.id
+    principalId: reference(identityName_resource.id, '2018-11-30').principalId
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:99]: use the bicep syntax here, e.g. ```msi.properties.principalId```

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  identity: {
+    type: 'UserAssigned'
+    userAssignedIdentities: {
+      '${identityName_resource.id}': {}
+    }
+  }
+  properties: {
+    buildTimeoutInMinutes: 60
+    vmProfile: {
+      vmSize: 'Standard_D1_v2'
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:149]: move this to V3 series

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    distribute: [
+      {
+        type: 'SharedImage'
+        galleryImageId: imageGalleryName_imageDefinition_imageDefinitionName.id
+        runOutputName: runOutputName
+        replicationRegions: replicationRegions
+      }
+    ]
+  }
+  dependsOn: [
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:185]: remove dependsOn (should be automatically added)

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+  location: location
+  kind: 'AzurePowerShell'
+  identity: {
+    type: 'UserAssigned'
+    userAssignedIdentities: {
+      '${identityName_resource.id}': {}
+    }
+  }
+  properties: {
+    forceUpdateTag: 'newGuid()'
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:201]: this isn't processed as an expresssion, you have it in quotes so it's just a string...  If you want to force this to run every time, you have to use newGuid() in a parameter's defaultValue

File: demos/imagebuilder-windowsbaseline/main.bicep [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    azPowerShellVersion: '6.2'
+    scriptContent: 'Invoke-AzResourceAction -ResourceName "${imageTemplateName}" -ResourceGroupName "${resourceGroup().name}" -ResourceType "Microsoft.VirtualMachineImages/imageTemplates" -ApiVersion "2020-02-14" -Action Run -Force'
+    timeout: 'PT1H'
+    cleanupPreference: 'OnSuccess'
+    retentionInterval: 'P1D'
+  }
+  dependsOn: [
```
Comment (bmoore-msft) [inline demos/imagebuilder-windowsbaseline/main.bicep:208]: move this above properties

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: demos/imagebuilder-windowsbaseline/README.md (hunk 1 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/README.md:49]: ack

File: demos/imagebuilder-windowsbaseline/main.bicep (hunk 2 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/main.bicep:21]: ack

File: demos/imagebuilder-windowsbaseline/main.bicep (hunk 3 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/main.bicep:41]: ack

File: demos/imagebuilder-windowsbaseline/main.bicep (hunk 4 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/main.bicep:69]: investigating

File: demos/imagebuilder-windowsbaseline/main.bicep (hunk 7 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/main.bicep:149]: ack

File: demos/imagebuilder-windowsbaseline/main.bicep (hunk 8 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/main.bicep:185]: ack

File: demos/imagebuilder-windowsbaseline/main.bicep (hunk 9 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/main.bicep:201]: ack

File: demos/imagebuilder-windowsbaseline/main.bicep (hunk 10 above)
Comment (mgreenegit) [reply demos/imagebuilder-windowsbaseline/main.bicep:208]: ack

**Comment by mgreenegit** [comment]
@bmoore-msft thanks for the review. :shipit:

**Comment by mgreenegit** [comment]
#sign-off

**Pull request merged**